codex/blockexchange/engine/pendingblocks.nim

```diff
@@ -36,12 +43,121 @@ type
     inFlight*: bool
     startTime*: int64
 
+  LeafReq* = object
+    case delivered*: bool
+    of false:
+      handle*: Future[Block]
+      inFlight*: bool
+    of true:
+      leaf: MultiHash
+      blkCid*: Cid
+
+  TreeReq* = ref object
+    leaves*: Table[Natural, LeafReq]
+    deliveredCount*: Natural
+    leavesCount*: ?Natural
+    treeRoot*: MultiHash
+    treeCid*: Cid
+
+  TreeHandler* = proc(tree: MerkleTree): Future[void] {.gcsafe.}
+
   PendingBlocksManager* = ref object of RootObj
     blocks*: Table[Cid, BlockReq] # pending Block requests
+    trees*: Table[Cid, TreeReq]
+    onTree*: TreeHandler
 
 proc updatePendingBlockGauge(p: PendingBlocksManager) =
   codexBlockExchangePendingBlockRequests.set(p.blocks.len.int64)
 
+type 
+  BlockHandleOrCid = object
+    case resolved*: bool
+    of true:
+      cid*: Cid
+    else:
+      handle*: Future[Block]
+
+proc buildTree(treeReq: TreeReq): ?!MerkleTree =
+  trace "Building a merkle tree from leaves", treeCid = treeReq.treeCid, leavesCount = treeReq.leavesCount
+
+  without leavesCount =? treeReq.leavesCount:
+    return failure("Leaves count is none, cannot build a tree")
+
+  var builder = ? MerkleTreeBuilder.init(treeReq.treeRoot.mcodec)
+  for i in 0..<leavesCount:
+    treeReq.leaves.withValue(i, leafReq):
+      if leafReq.delivered:
+        ? builder.addLeaf(leafReq.leaf)
+      else:
+        return failure("Expected all leaves to be delivered but leaf with index " & $i & " was not")
+    do:
+      return failure("Missing a leaf with index " & $i)
+
+  let tree = ? builder.build()
+
+  if tree.root != treeReq.treeRoot:
+    return failure("Reconstructed tree root doesn't match the original tree root, tree cid is " & $treeReq.treeCid)
+ 
+  return success(tree)
+
+proc checkIfAllDelivered(p: PendingBlocksManager, treeReq: TreeReq): void =
+  if treeReq.deliveredCount.some == treeReq.leavesCount:
+    without tree =? buildTree(treeReq), err:
+      error "Error building a tree", msg = err.msg
+      p.trees.del(treeReq.treeCid)
+      return
+    p.trees.del(treeReq.treeCid)
+    try:
+      asyncSpawn p.onTree(tree)
```

Comment:
Righ, I will change the log message, but I have to leave the `try/exept` because otherwise I get a compilation error:
> nim-codex/codex/blockexchange/engine/pendingblocks.nim(111, 24) Error: Exception can raise an unlisted exception: Exception

Reply:
Interesting, this should be handled elsewhere then, at the point where the exception is being initiated. The problem is that this isn't really doing anything at all and the correct way of handling the base `Exception` (which should never be raised explicitely), is to convert it to a `Defect` and re-raise. So we need to do some diggeing to determine what exactly is trying to raise the `Exception`.

Reply:
You can try to mark the `TreeHandler` type with an explicit `{.raises:[].}`

I have seen that sometimes that isn't enough, and the compiler mistakenly thinks that a closure can raise an `Exception`.  Then the only way to work around it is to add a `try:except` block, and raise a `Defect`.